engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java

```diff
@@ -173,37 +175,44 @@ public void execute(
           SourceNotFoundException,
           UnsupportedMessageException,
           UnsupportedTypeException {
-    SourceSection src = call.getFunction().getSourceSection();
-    if (src == null) {
-      throw new SourceNotFoundException(call.getFunction().getName());
-    }
-
-    var callbacks =
-        new ExecutionCallbacks(
-            visualizationHolder,
-            nextExecutionItem,
-            cache,
-            methodCallsCache,
-            syncState,
-            expressionExecutionState,
-            onCachedCallback,
-            onComputedCallback,
-            funCallCallback,
-            onExecutedVisualizationCallback,
-            this.context.isProgressReportEnabled() ? onComputedCallback : null);
-    Optional<EventBinding<ExecutionEventNodeFactory>> eventNodeFactory =
-        idExecutionInstrument.map(
-            service ->
-                service.bind(module, call.getFunction().getCallTarget(), callbacks, this.timer));
-
-    Object p = context.getThreadManager().enter();
-    try {
-      var callFn = Function.fullyApplied(execute.getCallTarget(), substituteMissingArguments(call));
-      RunStateNode.getUncached().execute(null, cacheKey(), cache, callFn);
-    } finally {
-      context.getThreadManager().leave(p);
-      eventNodeFactory.ifPresent(EventBinding::dispose);
-    }
+    var pending =
+        submitExecution(
+            () -> {
+              SourceSection src = call.getFunction().getSourceSection();
+              if (src == null) {
+                throw new SourceNotFoundException(call.getFunction().getName());
+              }
+
+              var callbacks =
+                  new ExecutionCallbacks(
+                      visualizationHolder,
+                      nextExecutionItem,
+                      cache,
+                      methodCallsCache,
+                      syncState,
+                      expressionExecutionState,
+                      onCachedCallback,
+                      onComputedCallback,
+                      funCallCallback,
+                      onExecutedVisualizationCallback,
+                      this.context.isProgressReportEnabled() ? onComputedCallback : null);
+              Optional<EventBinding<ExecutionEventNodeFactory>> eventNodeFactory =
+                  idExecutionInstrument.map(
+                      service ->
+                          service.bind(
+                              module, call.getFunction().getCallTarget(), callbacks, this.timer));
+
+              try {
+                var callFn =
+                    Function.fullyApplied(
+                        execute.getCallTarget(), substituteMissingArguments(call));
+                RunStateNode.getUncached().execute(null, cacheKey(), cache, callFn);
+              } finally {
+                eventNodeFactory.ifPresent(EventBinding::dispose);
+              }
+              return null;
+            });
+    resultOf(pending);
```

Comment:
- `ExecutionService` keeps compatibility with previous **blocking behavior**
- even the _Enso code_ computation is performed in _guest thread_
    - it has to to address https://github.com/enso-org/enso/pull/12500#issuecomment-2752028304
- e.g. the code is `submit`ed to `ThreadManager` for execution
- `Future` is returned
- but all methods in `ExecutionService` wait for such `Future`s to be finished
- these methods could be changed to return `Future` in subsequent PRs
- that would help with #10525
- e.g. this PR **opens the door** for _parallel evaluation of visualizations_